In [27]:
# WANT: Proj SOG

# Model:

# - SOG/GP TY
# - SOG/GP L3-L5
# - Opp Team Allowed SOG/GP TY
# - Opp Team Allowed SOG/GP L3-L5

# Additional:

# - Any Starting Opp D out
# - Starting Opp G out
# - Back to back games
# - Home vs Away
# - Penalties
# - 


import pandas as pd
fullPageData = pd.read_html('https://www.hockey-reference.com/teams/CGY/')
teamData = fullPageData[1]
rosterData = fullPageData[4]
rosterData.columns = rosterData.columns.droplevel()
print(rosterData.loc[:,["Player", "GP","S"]])

                Player  GP     S
0        Tyler Toffoli  57   175
1       Elias Lindholm  55   118
2          Nazem Kadri  57   189
3      Mikael Backlund  57   181
4   Jonathan Huberdeau  54    88
5     Rasmus Andersson  54   117
6          Dillon Dube  57   104
7    Andrew Mangiapane  57   128
8        Blake Coleman  57   129
9         Noah Hanifin  56   126
10        Adam Ruzicka  41    59
11    MacKenzie Weegar  56   100
12        Trevor Lewis  57   115
13      Nikita Zadorov  57    89
14         Milan Lucic  54    52
15         Chris Tanev  47    45
16       Michael Stone  46    86
17       Brett Ritchie  34    46
18      Dennis Gilbert  13     8
19       Connor Mackey  10     8
20        Walker Duehr   8    12
21     Jakob Pelletier  11    21
22     Jacob Markstrom  38     0
23        Kevin Rooney  17    14
24       Nick DeSimone   4     4
25    Matthew Phillips   2     1
26       Daniel Vladar  21     0
27       Radim Zohorna   8     3
28          Team Total  57  2018


In [28]:
SGPData = rosterData.loc[:,["Player", "GP","S"]]
SGPData['S/GP'] = SGPData['S']/SGPData['GP']
SGPData.drop(SGPData.tail(1).index,inplace=True)
SGPData['S/GP'] = pd.to_numeric(SGPData['S/GP'])
SGPData = SGPData.loc[SGPData['S/GP'] >= 1.5]
SGPData

Player  GP    S      S/GP
0        Tyler Toffoli  57  175  3.070175
1       Elias Lindholm  55  118  2.145455
2          Nazem Kadri  57  189  3.315789
3      Mikael Backlund  57  181  3.175439
4   Jonathan Huberdeau  54   88  1.629630
5     Rasmus Andersson  54  117  2.166667
6          Dillon Dube  57  104  1.824561
7    Andrew Mangiapane  57  128  2.245614
8        Blake Coleman  57  129  2.263158
9         Noah Hanifin  56  126  2.250000
11    MacKenzie Weegar  56  100  1.785714
12        Trevor Lewis  57  115  2.017544
13      Nikita Zadorov  57   89  1.561404
16       Michael Stone  46   86  1.869565
20        Walker Duehr   8   12  1.500000
21     Jakob Pelletier  11   21  1.909091

In [30]:
import numpy as np

playerL5AVG = []
for index, row in SGPData.iterrows():
    player = row['Player']
    url = 'https://www.hockey-reference.com/players/'
    nameSplit = player.split()
    url += nameSplit[1][0]
    url += '/'
    url += nameSplit[1][:5]
    url += nameSplit[0][:2]
    url += '01.html'
    url = url.lower()
    playerL5 = pd.read_html(url)
    playerL5 = playerL5[0]
    playerL5.columns = playerL5.columns.droplevel()
    playerL5.drop(playerL5.tail(1).index,inplace=True)
    playerL5 = pd.to_numeric(playerL5.loc[:,"S"])
    if (playerL5.mean() > 6):
        playerL5AVG.append(np.nan)
    else:
        playerL5AVG.append(playerL5.mean())



SGPData["S/GP L5"] = playerL5AVG
SGPData

Player  GP    S      S/GP  S/GP L5
0        Tyler Toffoli  57  175  3.070175      3.4
1       Elias Lindholm  55  118  2.145455      1.8
2          Nazem Kadri  57  189  3.315789      4.0
3      Mikael Backlund  57  181  3.175439      4.6
4   Jonathan Huberdeau  54   88  1.629630      1.8
5     Rasmus Andersson  54  117  2.166667      2.4
6          Dillon Dube  57  104  1.824561      2.6
7    Andrew Mangiapane  57  128  2.245614      3.0
8        Blake Coleman  57  129  2.263158      0.4
9         Noah Hanifin  56  126  2.250000      1.8
11    MacKenzie Weegar  56  100  1.785714      2.0
12        Trevor Lewis  57  115  2.017544      1.2
13      Nikita Zadorov  57   89  1.561404      1.6
16       Michael Stone  46   86  1.869565      2.0
20        Walker Duehr   8   12  1.500000      1.6
21     Jakob Pelletier  11   21  1.909091      2.2

Opponents Allowed SOG Avg

Example : 35.23

Avg SOG

Example: 35.4

Toffoli Avg:

~ 8.7% of the SOG for team

8.7% of 35.23 = Proj 3.065 SOG

Toffoli AVG L5:

~ 9.6%

9.6% of 35.23 = 3.382 SOG

AVG OF ALL SOG AVGS:

3.382 + 3.065 + 3.070 + 3.4 = Aggregate project SOG 3.229